### ISSUE TITLE
Connection is closed after each command.

### ISSUE THREAD

**Issue by florentbr** [mozilla/geckodriver#713]
While investigating on performance issues, I noticed that the geckodriver closes the http/tcp connection after each executed command.
The connection is dropped even if the header 'Connection: keep-alive' is set. Therefore, the client is forced to reconnect each time.

According to the specification, the connection should be established only at the beginning:

https://w3c.github.io/webdriver/webdriver-spec.html#processing-model

## Expected:
The TCP connection should remain opened until it is terminated by the client or until timeout.

## Firefox Version
Nightly 55.0a1

## Platform
Windows 10

## Steps to reproduce -

```python
import socket, sys, re

sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
sock.connect(('127.0.0.1', 4444))

request = '\r\n'.join(( \
  'POST /session HTTP/1.1',
  'Host: 127.0.0.1:4444',
  'Connection: keep-alive',
  'Accept: text/plain, application/json',
  'Cache-Control: no-store,no-cache'
  'Content-Type: application/json; charset=utf-8',
  'Content-Length: 130',
  '',
  '{"desiredCapabilities":{"pageLoadStrategy":"normal","unexpectedAlertBehaviour":"ignore","browserName":"firefox","platform":"ANY"}}'
))

sock.sendall(request.encode('utf-8'))
response = sock.recv(65000).decode('utf-8')

# HTTP/1.1 200 OK
# Connection: close
# Content-Type: application/json; charset=utf-8
# Cache-Control: no-cache
# Content-Length: 479
# Date: Wed, 03 May 2017 02:26:03 GMT
# 
# {"value": {"sessionId":"12abc5e2-18dd-4e12-bb39-17f157b1ffdf","capabilities":{"acceptInsec...

sessionId = re.search(r'sessionId":"([^"]+)', response).group(1)

request = '\r\n'.join((\
  'POST /session/%s/url HTTP/1.1' % sessionId,
  'Host: 127.0.0.1:4444',
  'Connection: keep-alive',
  'Accept: text/plain, application/json',
  'Cache-Control: no-store,no-cache'
  'Content-Type: application/json; charset=utf-8',
  'Content-Length: 24',
  '',
  '{"url": "about:mozilla"}'
))

sock.sendall(request.encode('utf-8'))
response = sock.recv(65000).decode('utf-8')

ConnectionAbortedError: [WinError 10053] An established connection was aborted by the software in your host machine

```

**Comment by andreastt** [comment]
geckodriver uses an HTTP API, and trying to use that through a low-level TCP socket is going to cause you grief as you need to handle the prospect that the socket can be closed at any time and other details relevant to HTTP.

Now as far as I know, this won’t work because the [webdriver](https://crates.io/crates/webdriver) crate doesn’t do Keep-Alive connections (see https://github.com/mozilla/webdriver-rust/blob/d7fe400fa903b752b3c90c2c7e80ae9d5719e16b/src/server.rs#L253).

The WebDriver specification doesn’t explicitly say that the remote end needs to support Keep-Alive connections. All it says about the remote end is that it needs to be [‘HTTP compliant’](https://w3c.github.io/webdriver/webdriver-spec.html#dfn-http-compliant), which doesn’t necessarily entail supporting Keep-Alive.

HTTP is transactional and stateless by design, and the normal way of operating HTTP is by making one connection per request. With HTTP 1.1 it is implied that connections are persistent and it is therefore possible we’re failing to return a header describing that the connection _isn’t_.

**Comment by florentbr** [comment]
The code I posted is just here to demonstrate the issue and I'm not using it and I don't intend to use it.
I noticed the issue while watching the logs through a proxy to check if the packets were efficiently posted.

I agree that the specification doesn't explicitly say that the remote end needs to support Keep-Alive.
However it implicitly says that the connection should be persistent (note that the HTTP specs doesn't impose the `Keep-Alive` header for a persistent connection).

From chapter 6.3 :

> After such a connection has been established, a remote end must run the following steps:
>>
>>  1. Read bytes from the connection until a complete HTTP request can be constructed from the data. Let request be a request constructed from the received data, according to the requirements of [RFC7230].
>>  ...
>>  10. Send a response with status 200 and response data.
>>  11. Jump to step 1.


The Chrome driver does comply to this specified behavior, but the geckodriver doesn't and waits time on handshakes for no reason.

**Comment by andreastt** [comment]
I disagree with you that the WebDriver specification implicitly says that. The wording around reading enough data off the connection until an HTTP request can be constructed is intentionally fuzzy because this isn’t described anywhere else and it isn’t the right place to define it. But that quote doesn’t imply that said connection is reused from reading the preceding request: it only talks in very loose terms about getting to the point where you have an HTTP request.

If I remember correctly, there were issues with existing Selenium clients that caused us to explicitly disable Keep-Alive connections in the HTTPD, see https://github.com/mozilla/webdriver-rust/commit/b614a0f45edf60fbc8931688a818edfdc4892cd7:

> ```
> Don't use keepalive in the http server.
>
> This isn't supported by all clients (e.g. python's httplib)
> and so can cause the connection to be unexpectedly dropped.
> A better long-term solution is likely needed for this case,
> for example allowing the caller to decide if it is enabled.
> ```

By “better solution”, it probably ought to look at the HTTP request and only disable Keep-Alive if the HTTP/1.1 request explicitly does not want it, which should be easy enough to instruct httplib to do. The relevant client using httplib in this case is the WPT WebDriver client found in https://github.com/w3c/web-platform-tests/tree/master/tools/webdriver.

Out of interest, in what sort of ballpark are the performance penalties you are seeing? My educated guess would be that there are far worse time thieves than TCP connections and HTTP handshakes that slows down WebDriver.

**Comment by cgoldberg** [comment]
> This isn't supported by all clients (e.g. python's httplib)

fwiw, httplib does support keep-alives.  It's just a header you set.

**Comment by andreastt** [comment]
> fwiw, httplib does support keep-alives. It's just a header you set.

It’s not clear to me what this means. In HTTP 1.1, persistent connections (aka `Keep-Alive`) is implicit. That means the server needs to respond explicitly to say it doesn’t support it, if that’s the case.

If httplib needs to have `Keep-Alive` set as a header explicitly when making requests, that means it implements HTTP 1.0, where that was required for establishing a persistent connection.

**Comment by cgoldberg** [comment]
> If httplib needs to have Keep-Alive set as a header explicitly

httplib supports both http 1.0 and 1.1.
if using 1.0, "Connection: Keep-Alive" must be explicitly set, otherwise it will send a "Connection: Close" header.  When speaking 1.1,  httplib will keep the socket connection open by default.

so... python's httplib does allow keep-alives, regardless of protocol version.  sorry if my previous comment was misleading.

**Comment by florentbr** [comment]
@andreastt, my interpretation is "_After such a connection has been established, execute one or more commands (Steps 1 to 11)_", which is why I say it implies a persistent connection.

If a client is unable to recover from an unexpected drop of connection or doesn't want a persistent connection, it should add the header 'Connection: close' and close the connection after sending the response.

From the HTTP 1.1 protocol:

> An HTTP/1.1 server MAY assume that a HTTP/1.1 client intends to maintain a persistent connection unless a Connection header including the connection-token "close" was sent in the request. If the server chooses to close the connection immediately after sending the response, it SHOULD send a Connection header including the connection-token close. 

The penalty related to the persistence is probably insignificant. I can't measure it since the geckodriver 
 doesn't allow it.

> Out of interest, in what sort of ballpark are the performance penalties you are seeing? My educated guess would be that there are far worse time thieves than TCP connections and HTTP handshakes that slows down WebDriver.

I went further and wrote a test to send some commands directly to Firefox/marionette via TCP. I measured that going through the geckodriver only adds a few milliseconds to each command.

So feel free to close this issue if you think that disabling the persistence of the connection doesn't goes against the specification.

Anyway I think I narrowed the performance issue down to the warping/unwarping of elements in marionette.
Here's an example to observe the performance issue if you are interested:

```python
driver.get('https://www.w3.org/TR/webdriver/#protocol')

elems = driver.find_elements_by_css_selector('p')  # 2583 ms / warps 1371 elements

driver.refresh()
elems = driver.find_elements_by_css_selector('p')  # 7736 ms / warps 1371 elements

driver.refresh()
elems = driver.find_elements_by_css_selector('p')  # 13864 ms / warps 1371 elements

driver.execute_script("return 0;")                 # 61 ms  / unwarps 0 elements
driver.execute_script("return 0;", elems)          # 5896 ms  / unwarps 1371 elements
```

**Comment by andreastt** [comment]
Also sprach Florent Breheret:

> The penalty related to the persistence is probably insignificant. I
> can't measure it since the geckodriver doesn't allow it.

You could probably run an intentionally expensive test, such as
executing a few thousand empty scripts or getting the window title,
against one build where persistent connections are enabled and one in
which it isn’t.

I’m trying to judge if fixing this has any practical benefit.

>> Out of interest, in what sort of ballpark are the performance
>> penalties you are seeing? My educated guess would be that there are
>> far worse time thieves than TCP connections and HTTP handshakes that
>> slows down WebDriver.
> 
> I went further and wrote a test to send some commands directly
> to Firefox/marionette via TCP. I measured that going through the
> geckodriver only adds a few milliseconds to each command.

But that’s not really a useful comparison.  What we’re concerned
with is what kind of impact having an HTTPD that doesn’t offer
persistent connections has.

> So feel free to close this issue if you think that disabling the
> persistence of the connection doesn't goes against the specification.

I didn’t say exactly that.  It is in line with the specification to
offer an “HTTP compliant server”, but because the language around
this is somewhat intentionally vague, my argument was that we were not
_incompliant_ by not offering persistent connections.

I want to enable persistent connections in geckodriver, but we need to
be sure that the existing corpus of clients can handle it.  Judging by
@cgoldberg’s feedback it seems that httplib can handle it, but I need
to confirm that in practice.

> Anyway I think I narrowed the performance issue down to the
> warping/unwarping of elements in marionette.

That looks quite bad, but do you mind filing a new issue about this so
we don’t bring this issue off-topic?

**Comment by florentbr** [comment]
@andreastt, I don't have the environment to build a modified version of the geckodriver.

My hypothesis was that if the time to go through the geckodriver is insignificant, then the time to open/close the connection with the geckodriver is also insignificant, which is the case.

> I’m trying to judge if fixing this has any practical benefit.

 Pros:
 * Comply to the default behavior of HTTP 1.1 (rfc7230: HTTP implementations SHOULD support persistent connections).
 * Comply to the webdriver specifications (might be subject to interpretation).
 * The client has the option to use a persistent connection or a non persistent connection.
 
Cons:
 * Low benefit in terms of performance (insignificant on a use-case).
 * Probably requires to update the code of each client so that:
    The client no longer sends the header "Connection: keep-alive" for a persistent connection (HTTP/1.0).
    The client sends the header "Connection: close" if it requires a non persistent connection (HTTP/1.1).
    The client considers the connection persistent by default unless the server returns the header "Connection: close".
    If a client decides to keep up with a persistent connection, it should also implement a retry mechanism in case the connection is closed prematurely.
  
Even though allowing the persistence might seems to be the right thing to do, it probably doesn't worth the change in regard to the possible impact on each client.

Note that is a previous comment, I mentioned that the chromedriver allows a persistent connection.
Well, I found out that it is half right/wrong. While the connection remains opened, the driver always returns the header "Connection: close" to instruct the client to cease sending requests on that connection.

**Comment by AutomatedTester** [comment]
> Note that is a previous comment, I mentioned that the chromedriver allows a persistent connection.
Well, I found out that it is half right/wrong. While the connection remains opened, the driver always returns the header "Connection: close" to instruct the client to cease sending requests on that connection.

Actually... None of the Drivers (InternetExplorerDriver, ChromeDriver or GeckoDriver) handle HTTP/1.1. I have started playing with switching on keep-alive. The unfortunate thing is the unexpected drops in connections. The Selenium clients don't always rerequest, and rerequesting could be fraught with interesting side effects like 2nd clicks on same element...

**Comment by AutomatedTester** [comment]
I do get the argument for trying to prevent problem of running out of ephemeral ports, especially on Windows.

**Comment by florentbr** [comment]
Do we have any clues on whether the unexpected drop of connection is inherent to the server or to the OS?
I don't see why this happens since the connection is local and not exposed to a network failure.

**Comment by AutomatedTester** [comment]
I've raised https://bugzilla.mozilla.org/show_bug.cgi?id=1468398

**Comment by whimboo** [comment]
This is fixed in the 0.21.0 release of geckodriver.
